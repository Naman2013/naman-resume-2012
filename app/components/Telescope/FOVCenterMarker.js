import React from 'react';
import PropTypes from 'prop-types';
import isFinite from 'lodash/isFinite';

const COVER_PERCENTAGE = 0.2;
const DASHED_RECT_PERCENTAGE = 0.25;
const DASHED_PERCENTAGE = 0.1;

const FOVCenterMarker = ({ gridDimension, gridWidth, canvasWidth }) => {
  if (!isFinite(COVER_PERCENTAGE)) return null;

  const smallRectDimension = (gridWidth * gridDimension);
  const largeRectWidth = ((smallRectDimension * DASHED_RECT_PERCENTAGE) + smallRectDimension);
  const smallRectX = ((canvasWidth / 2) - (smallRectDimension / 2));
  const largeRectX = ((canvasWidth / 2) - (largeRectWidth / 2));
  const Y = ((canvasWidth / 2) - (smallRectDimension / 2));
  const strokeDashArray = (smallRectDimension * DASHED_PERCENTAGE);
  return (
    <g>
      <rect
        x={largeRectX}
        y={Y}
        width={largeRectWidth}
        height={smallRectDimension}
        stroke="green"
        strokeDasharray={`${strokeDashArray}, ${strokeDashArray}`}
        fill="none"
      />
      <rect
        x={smallRectX}
        y={Y}
        width={smallRectDimension}
        height={smallRectDimension}
        stroke="green"
        fill="none"
      />
    </g>
  );
};

FOVCenterMarker.propTypes = {
  gridDimension: PropTypes.number.isRequired,
  gridWidth: PropTypes.number.isRequired,
  canvasWidth: PropTypes.number.isRequired,
};

export default FOVCenterMarker;
