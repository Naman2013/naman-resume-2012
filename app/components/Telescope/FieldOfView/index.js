import FieldOfView from './FieldOfView';

export default FieldOfView;
