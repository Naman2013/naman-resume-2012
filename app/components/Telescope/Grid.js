import React from 'react';
import PropTypes from 'prop-types';

const Circle = ({ x, y }) => (
  <circle fill="#319fff" cx={x} cy={y} r="2" />
);

function generateGrid(count, dimension) {
  const grid = [];
  const GRID_COUNT = (count * 2);
  const SPLIT = (count / 2);
  const SPACING = (dimension / count);

  for (let i = 0; i < count; i += 1) {
    grid.push(<Circle x={i} y={i} />);
  }

  return grid;
}

const Grid = ({ count, dimension }) => (
  <g>
    {generateGrid(count, dimension)}
  </g>
);

Grid.propTypes = {
  count: PropTypes.number,
  dimension: PropTypes.number,
};

Grid.defaultProps = {
  count: 50,
  dimension: 50,
};

export default Grid;
