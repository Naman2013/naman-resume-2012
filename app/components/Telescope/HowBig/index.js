import HowBig from './HowBig';

export default HowBig;
