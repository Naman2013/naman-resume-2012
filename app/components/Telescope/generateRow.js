import React from 'react';
import ROW_CONFIG from './rowConfigurationEnum';
import Polyline from '../SVG/Polyline';

function warnClient() {
  console.warn('Invalid row configuration provided');
}

export function generateRow(
  dimension = 0,
  count = 0,
  rowConfiguration = ROW_CONFIG.TOP,
  style = {},
) {
  if (!(rowConfiguration instanceof ROW_CONFIG)) {
    warnClient();
    return [];
  }
  if (!dimension) { return []; }

  const row = [];
  const COUNT = count;
  const TICKS_PER_SIDE = (count / 2);
  const SPACING = (dimension / count);
  const MID_POINT = (dimension / 2);
  const CENTER_TICK_LENGTH = 16;
  const CENTER_TICK_THICKNESS = 3;
  const LARGE_TICK_LENGTH = 12;
  const LARGE_TICK_THICKNESS = 2;
  const SHORT_TICK_LENGTH = (LARGE_TICK_LENGTH / 2);
  const SHORT_TICK_THICKNESS = (LARGE_TICK_THICKNESS / 2);

  let ACCUMULATOR = 0;

  let LEFT_ACCUMULATOR = MID_POINT;
  let RIGHT_ACCUMULATOR = MID_POINT;

  for (let i = 0; i <= COUNT; i += 1) {
    const isLongTick = (i % 5) === 0;

    let x1;
    let y1;
    let x2;
    let y2;
    let tickThickness;

    ACCUMULATOR += SPACING;

    switch (rowConfiguration) {
      case ROW_CONFIG.TOP:
        if (i === 0) {
          x1 = MID_POINT;
          y1 = 0;
          x2 = MID_POINT;
          y2 = CENTER_TICK_LENGTH;
          tickThickness = CENTER_TICK_THICKNESS;
        } else if (i <= TICKS_PER_SIDE) {
          x1 = LEFT_ACCUMULATOR;
          y1 = 0;
          x2 = LEFT_ACCUMULATOR;
          y2 = (isLongTick) ? LARGE_TICK_LENGTH : SHORT_TICK_LENGTH;
          tickThickness = (isLongTick) ? LARGE_TICK_THICKNESS : SHORT_TICK_THICKNESS;

          LEFT_ACCUMULATOR -= SPACING;
        } else {
          x1 = RIGHT_ACCUMULATOR;
          y1 = 0;
          x2 = RIGHT_ACCUMULATOR;
          y2 = (isLongTick) ? LARGE_TICK_LENGTH : SHORT_TICK_LENGTH;
          tickThickness = (isLongTick) ? LARGE_TICK_THICKNESS : SHORT_TICK_THICKNESS;

          RIGHT_ACCUMULATOR += SPACING;
        }
        break;
      case ROW_CONFIG.BOTTOM:
        x1 = ACCUMULATOR;
        y1 = dimension;
        x2 = ACCUMULATOR;
        y2 = (isLongTick) ? (dimension - LARGE_TICK_LENGTH) : (dimension - SHORT_TICK_LENGTH);
        tickThickness = (isLongTick) ? LARGE_TICK_THICKNESS : SHORT_TICK_THICKNESS;

        if (i === 0) {
          x1 = MID_POINT;
          y1 = dimension;
          x2 = MID_POINT;
          y2 = (dimension - CENTER_TICK_LENGTH);
          tickThickness = CENTER_TICK_THICKNESS;
        } else if (i <= TICKS_PER_SIDE) {
          x1 = LEFT_ACCUMULATOR;
          y1 = dimension;
          x2 = LEFT_ACCUMULATOR;
          y2 = (isLongTick) ? (dimension - LARGE_TICK_LENGTH) : (dimension - SHORT_TICK_LENGTH);
          tickThickness = (isLongTick) ? LARGE_TICK_THICKNESS : SHORT_TICK_THICKNESS;

          LEFT_ACCUMULATOR -= SPACING;
        } else {
          x1 = RIGHT_ACCUMULATOR;
          y1 = dimension;
          x2 = RIGHT_ACCUMULATOR;
          y2 = (isLongTick) ? (dimension - LARGE_TICK_LENGTH) : (dimension - SHORT_TICK_LENGTH);
          tickThickness = (isLongTick) ? LARGE_TICK_THICKNESS : SHORT_TICK_THICKNESS;

          RIGHT_ACCUMULATOR += SPACING;
        }
        break;
      case ROW_CONFIG.LEFT:
        x1 = 0;
        y1 = ACCUMULATOR;
        x2 = (isLongTick) ? LARGE_TICK_LENGTH : SHORT_TICK_LENGTH;
        y2 = ACCUMULATOR;
        tickThickness = (isLongTick) ? LARGE_TICK_THICKNESS : SHORT_TICK_THICKNESS;
        break;
      case ROW_CONFIG.RIGHT:
        x1 = dimension;
        y1 = ACCUMULATOR;
        x2 = (isLongTick) ? (dimension - LARGE_TICK_LENGTH) : (dimension - SHORT_TICK_LENGTH);
        y2 = ACCUMULATOR;
        tickThickness = (isLongTick) ? LARGE_TICK_THICKNESS : SHORT_TICK_THICKNESS;
        break;
      default:
        warnClient();
        break;
    }

    const polylineAttributes = {
      points: `${x1},${y1} ${x2},${y2}`,
      strokeWidth: tickThickness,
    };

    row.push(<Polyline {...polylineAttributes} {...style} />);
  }

  return row;
}
