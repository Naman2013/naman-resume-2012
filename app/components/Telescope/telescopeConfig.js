import first from 'lodash/first';

const telescopeConfig = {
  '1ff72faa-7909-11e6-a635-0eb2b1774883': {
    key: 'CANARY_ONE_HALF_METER',
    name: 'Canary One Half Meter',
    topName: 'Canary One',
    bottomName: 'Half Meter',
    instrumentID: '76de934d-7909-11e6-a635-0eb2b1774883',
    directionMarkerLengthArcMinutes: 3,
    FOV: {
      horizontal: 37,
      vertical: 37,
    },
    PORTAL: {
      horizontal: 37,
      vertical: 37,
    },
  },
  '2590c3fd-7909-11e6-a635-0eb2b1774883': {
    key: 'CANARY_TWO_WIDE_FIELD',
    name: 'Canary Two Wide Field',
    topName: 'Canary Two',
    bottomName: 'Wide-Field',
    instrumentID: '8032dfaf-7909-11e6-a635-0eb2b1774883',
    directionMarkerLengthArcMinutes: 3,
    FOV: {
      horizontal: 43,
      vertical: 43,
    },
    PORTAL: {
      horizontal: 43,
      vertical: 43,
    },
  },
  CANARY_TWO_ULTRA_WIDE_FIELD: {
    key: 'CANARY_TWO_ULTRA_WIDE_FIELD',
    name: 'Canary Two Ultra Wide Field',
    topName: 'Canary Two',
    bottomName: 'Ultra-Wide-Field',
    instrumentID: '88814408-7909-11e6-a635-0eb2b1774883',
    directionMarkerLengthArcMinutes: 6,
    FOV: {
      horizontal: 107,
      vertical: 72,
    },
    PORTAL: {
      horizontal: 72,
      vertical: 72,
    },
  },
  '3686b322-7909-11e6-a635-0eb2b1774883': {
    key: 'CANARY_THREE_DEEP_SKY',
    name: 'Canary Three Deep Sky',
    topName: 'Canary Three',
    bottomName: 'Deep Sky',
    instrumentID: '905f9d8d-7909-11e6-a635-0eb2b1774883',
    directionMarkerLengthArcMinutes: 6,
    FOV: {
      horizontal: 99,
      vertical: 75,
    },
    PORTAL: {
      horizontal: 75,
      vertical: 75,
    },
  },
  '3db71869-7909-11e6-a635-0eb2b1774883': {
    key: 'CANARY_FOUR_SOLAR_SYSTEM',
    name: 'Canary Four Solar System',
    topName: 'Canary Four',
    bottomName: 'Solar System',
    instrumentID: '97f58d52-7909-11e6-a635-0eb2b1774883',
    directionMarkerLengthArcMinutes: 1,
    FOV: {
      horizontal: 16,
      vertical: 12,
    },
    PORTAL: {
      horizontal: 12,
      vertical: 12,
    },
  },
  '54accc1d-7909-11e6-a635-0eb2b1774883': {
    key: 'CANARY_FIVE_SOLAR',
    name: 'Canary Five Solar',
    instrumentID: 'b0b33e71-7909-11e6-a635-0eb2b1774883',
    directionMarkerLengthArcMinutes: 2,
    FOV: {
      horizontal: 37,
      vertical: 21,
    },
    PORTAL: {
      horizontal: 21,
      vertical: 21,
    },
  },
  '4624887a-7909-11e6-a635-0eb2b1774883': {
    key: 'CHILE_ONE_WIDE_FIELD',
    name: 'Chile One Wide Field',
    topName: 'Chile One',
    bottomName: ' Wide-Field',
    instrumentID: 'cd31c5c9-7909-11e6-a635-0eb2b1774883',
    directionMarkerLengthArcMinutes: 2,
    FOV: {
      horizontal: 31,
      vertical: 21,
    },
    PORTAL: {
      horizontal: 21,
      vertical: 21,
    },
  },
  '6acfb8cd-7909-11e6-a635-0eb2b1774883': {
    key: 'CHILE_ONE_ULTRA_WIDE_FIELD',
    name: 'Chile One Ultra Wide Field',
    topName: 'Chile One',
    bottomName: 'Ultra-Wide-Field',
    instrumentID: 'd3ca2e16-7909-11e6-a635-0eb2b1774883',
    directionMarkerLengthArcMinutes: 5,
    FOV: {
      horizontal: 82,
      vertical: 60,
    },
    PORTAL: {
      horizontal: 60,
      vertical: 60,
    },
  },
  '8bc4567f-c9f4-11e6-86df-0eb2b1774883': {
    key: 'CHILE_TWO_WIDE_FIELD',
    name: 'Chile Two Wide Field',
    topName: 'Chile Two',
    bottomName: 'Wide-Field',
    instrumentID: '9cf72bb2-34ad-11e7-970a-0eb2b1774883',
    directionMarkerLengthArcMinutes: 3,
    FOV: {
      horizontal: 43,
      vertical: 43,
    },
    PORTAL: {
      horizontal: 43,
      vertical: 43,
    },
  },
};

export default telescopeConfig;

export function getTelescope(instrumentID) {
  const telescopeNames = Object.keys(telescopeConfig);
  const selectedTelescope = first(telescopeNames.filter(telescope => telescopeConfig[telescope].instrumentID === instrumentID));
  return telescopeConfig[selectedTelescope];
}
