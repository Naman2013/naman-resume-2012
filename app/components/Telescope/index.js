import Telescope from './Telescope';

export default Telescope;
