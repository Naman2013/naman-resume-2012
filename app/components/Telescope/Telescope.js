// @flow
import React, { PureComponent, Fragment } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Measure from 'react-measure';
import noop from 'lodash/noop';
import { setPreviousInstrument } from 'app/modules/starshare-camera/starshare-camera-actions';
import Fade from '../common/Fade';
import FadeSVG from '../common/Fade/FadeSVG';
import easingFunctions, { animateValues } from '../../utils/easingFunctions';
import Button from '../common/style/buttons/Button';

import TelescopeFrame from './TelescopeFrame';
import Mask from './Mask';
import Scale from './Scale';
import UnitText from './TelescopeFrame/UnitText';
import HowBig from './HowBig';

import { getTelescope } from './telescopeConfig';
import FieldOfView from './FieldOfView/FieldOfView';
import { Modal } from '../modal';

import { moodyBleu, romance } from '../../styles/variables/colors_tiles_v4';

const MAX_RESOLUTION = 250;
const MAX_DURATION = 10000;
const ZOOM_OUT_DURATION = MAX_DURATION / 2;
const MAX_FOV_FLIPS = 5;

const menuButtonTheme = {
  fontSize: '18px',
  borderWidth: '3px',
  background: 'black',
};

const inactiveButtonTheme = {
  ...menuButtonTheme,
  color: moodyBleu,
  borderColor: moodyBleu,
};

const activeButtonTheme = {
  ...menuButtonTheme,
  color: romance,
  borderColor: romance,
};

type TTelescope = {
  activeInstrumentID: string,
  previousInstrumentID: string | void,
  missionMetaData?: {
    missionTargetID?: number,
    referenceObjectScale?: number,
    domain?: string,
    targetObjectScale?: number,
    targetObjectURL?: string,
    targetObjectName?: string,
  },
  render?: Function,
  increment?: number,
  disableFullscreen?: boolean,
};

class Telescope extends PureComponent<TTelescope> {
  static defaultProps = {
    increment: 5,
    render: noop,
    missionMetaData: { missionTargetID: 0 },
  };

  state = {
    activeInstrumentID: this.props.activeInstrumentID,
    previousInstrumentID: this.props.previousInstrumentId,
    telescopeId: this.props.activeInstrumentID,
    timesFlippedInstrumentBorder: 0,
    isTransitioningTelescope: false,
    horizontalResolution: getTelescope(this.props.activeInstrumentID).FOV
      .horizontal,
    verticalResolution: getTelescope(this.props.activeInstrumentID).FOV
      .vertical,
    increment: this.props.increment,
    awaitingMission: this.props.missionMetaData.missionTargetID === 0,
    transitionScale: false,
    isMaskActive: false,
    isModalActive: false,
    isGridActive: true,
    showTitleMessage: false,
    transitionStrokeColor: 'aqua',
    showArrows: false,
    portalDimensions: {
      bottom: 0,
      height: 0,
      left: 0,
      right: 0,
      top: 0,
      width: 0,
      x: 0,
      y: 0,
    },
    radius: 0,
  };

  componentWillReceiveProps({
    activeInstrumentID,
    previousInstrumentId,
    missionMetaData,
  }) {
    this.props.setPreviousInstrument(activeInstrumentID);
    if (
      previousInstrumentId !== null &&
      previousInstrumentId !== activeInstrumentID
    ) {
      const tele = getTelescope(previousInstrumentId);
      this.setState(() => ({
        activeInstrumentID: previousInstrumentId,
        previousInstrumentID: activeInstrumentID,
        telescopeId: previousInstrumentId,
        horizontalResolution: tele.FOV.horizontal,
        verticalResolution:tele.FOV.vertical
      }));
      this.showTitleMessage();
    }

    if (missionMetaData.missionTargetID === 0) {
      this.setState(() => ({
        awaitingMission: true,
        transitionScale: false,
      }));
    }

    if (missionMetaData.missionTargetID !== 0) {
      this.setState(() => ({
        awaitingMission: false,
        transitionScale:
          missionMetaData.missionTargetID !==
          this.props.missionMetaData.missionTargetID,
      }));
    }
  }

  currentZoomInTransition = null;

  currentZoomOutTransition = null;

  doFOVTransitionInterval = null;

  showTitleMessage = () => {
    this.setState({ showTitleMessage: true }, () => {
      setTimeout(() => {
        this.transitionZoomOut();
      }, 3000);
    });
  };

  transitionZoomOut() {
    let remainingDuration = 0;

    this.setState(() => ({
      isTransitioningTelescope: true,
      showTitleMessage: false,
      showArrows: true,
    }));

    if (this.currentZoomOutTransition) {
      remainingDuration = this.currentZoomOutTransition
        .cancel()
        .getRemainingTime();
    }

    if (this.currentZoomInTransition) {
      remainingDuration =
        ZOOM_OUT_DURATION -
        this.currentZoomInTransition.cancel().getRemainingTime();
    }

    this.currentZoomInTransition = null;
    this.currentZoomOutTransition = null;

    this.currentZoomOutTransition = this.transitionTo(
      this.transitionPOV,
      {
        horizontal: MAX_RESOLUTION,
        vertical: MAX_RESOLUTION,
      },
      remainingDuration > 0 ? remainingDuration : ZOOM_OUT_DURATION
    );
  }

  transitionPOV() {
    this.setState({
      timesFlippedInstrumentBorder: 0,
      transitionStrokeColor: 'aqua',
      showArrows: false,
    });
    this.doFOVTransitionInterval = setInterval(() => {
      this.setState(prevState => {
        const {
          activeInstrumentID,
          previousInstrumentID,
          timesFlippedInstrumentBorder,
          transitionStrokeColor,
        } = prevState;

        if (timesFlippedInstrumentBorder >= MAX_FOV_FLIPS) {
          this.tearDownTransitionPOV();
          this.transitionZoomIn();
          return {};
        }

        const updatedFOVFlipState = {
          timesFlippedInstrumentBorder: timesFlippedInstrumentBorder + 1,
          activeInstrumentID:
            activeInstrumentID === this.state.activeInstrumentID
              ? previousInstrumentID
              : activeInstrumentID,
          previousInstrumentID:
            previousInstrumentID === this.state.previousInstrumentID
              ? activeInstrumentID
              : previousInstrumentID,
          transitionStrokeColor:
            transitionStrokeColor === 'aqua' ? '#FAD59A' : 'aqua',
          telescopeId:
            this.state.telescopeId === this.state.previousInstrumentID
              ? activeInstrumentID
              : previousInstrumentID,
        };
        return updatedFOVFlipState;
      });
    }, 500);
  }

  tearDownTransitionPOV() {
    if (this.doFOVTransitionInterval) {
      clearInterval(this.doFOVTransitionInterval);
    }
  }

  transitionZoomIn() {
    this.setState({ showArrows: true });
    const targetTelescope = getTelescope(this.state.activeInstrumentID);
    this.currentZoomInTransition = this.transitionTo(
      this.telescopeTransitionComplete,
      {
        horizontal: targetTelescope.PORTAL.horizontal,
        vertical: targetTelescope.PORTAL.vertical,
      }
    );
    this.setState({ transitionStrokeColor: 'aqua' });
  }

  telescopeTransitionComplete() {
    this.setState(() => ({ isTransitioningTelescope: false }));
  }

  transitionTo(
    onCompleteCallback,
    { horizontal, vertical },
    duration = ZOOM_OUT_DURATION
  ) {
    const { horizontalResolution, verticalResolution } = this.state;

    return animateValues(
      { hr: horizontalResolution, vr: verticalResolution },
      duration,
      {
        hr: horizontal,
        vr: vertical,
        onUpdate: values => {
          this.setState(() => ({
            horizontalResolution: values.hr,
            verticalResolution: values.vr,
          }));
        },
        onComplete: onCompleteCallback.bind(this),
        ease: easingFunctions.easeInOutQuint,
      }
    );
  }

  handlePortalResize = contentBox => {
    this.setState({ portalDimensions: { ...contentBox.bounds } });
  };

  handleCompleteHowBigAnimation = () => {
    this.setState(() => ({ transitionScale: false }));
  };

  onHideModal = () => {
    this.setState({ isModalActive: false });
  };

  render() {
    const {
      portalDimensions: { width, height },
      increment,
      horizontalResolution,
      verticalResolution,
      isTransitioningTelescope,
      activeInstrumentID,
      previousInstrumentID,
      transitionScale,
      awaitingMission,
      isMaskActive,
      isModalActive,
      isGridActive,
      radius,
      missionTitle,
    } = this.state;

    const { missionMetaData, disableFullscreen } = this.props;

    const activeInstrument = getTelescope(activeInstrumentID);
    console.log(horizontalResolution);
    const tickSpacing = width / horizontalResolution;
    const midPoint = width / 2;
    const arcMinuteLabelLetterSpacing = width * 0.03;

    return (
      <Measure bounds onResize={this.handlePortalResize}>
        {({ measureRef }) => (
          <div className="telescope">
            <div
              ref={measureRef}
              style={{
                backgroundColor: isTransitioningTelescope
                  ? 'black'
                  : 'transparent',
              }}
              className="portal"
            >
              {!isTransitioningTelescope && !this.state.showTitleMessage && (
                <div className="telescope-float-menu">
                  <Button
                    renderIcon={() => <i className="fa fa-table" />}
                    isActive={isGridActive}
                    onClickEvent={() => {
                      this.setState({ isGridActive: !isGridActive });
                    }}
                    theme={{
                      ...(isGridActive
                        ? activeButtonTheme
                        : inactiveButtonTheme),
                      marginBottom: '10px',
                    }}
                  />
                  <Button
                    renderIcon={() => <i className="fa fa-eye" />}
                    isActive={isMaskActive}
                    onClickEvent={() => {
                      this.setState({ isMaskActive: !isMaskActive });
                    }}
                    theme={{
                      ...(isMaskActive
                        ? activeButtonTheme
                        : inactiveButtonTheme),
                      marginBottom: '10px',
                    }}
                  />
                  {!disableFullscreen && (
                    <Button
                      renderIcon={() => <i className="fa fa-arrows-alt" />}
                      onClickEvent={() =>
                        this.setState({
                          isModalActive: true,
                          isMaskActive: false,
                        })
                      }
                      theme={inactiveButtonTheme}
                    />
                  )}
                </div>
              )}
              <Fade
                isHidden={
                  isTransitioningTelescope || this.state.showTitleMessage
                }
              >
                <div>
                  {this.props.render({ viewportHeight: width }, imageData => {
                    const { imageWidth, imageHeight, missionTitle } = imageData;

                    const radiusSize = (imageHeight * 0.65) / 2;

                    this.setState({
                      radius: radiusSize,
                      missionTitle,
                    });
                  })}
                </div>
              </Fade>
              <svg
                version="1.1"
                xmlns="http://www.w3.org/2000/svg"
                className="telescope-svg"
              >
                {/**
                 TODO:
                 move non-scale transition elements into a component to keep this more readable
                 */}
                <FadeSVG isHidden={transitionScale}>
                  <FadeSVG
                    isHidden={
                      isTransitioningTelescope || this.state.showTitleMessage
                    }
                  >
                    {isMaskActive && <Mask radius={radius} />}
                  </FadeSVG>
                  {(this.state.showTitleMessage ||
                    isTransitioningTelescope) && (
                    <g>
                      <rect
                        x="0"
                        y="0"
                        width={width}
                        height={height}
                        style={{ fill: 'black' }}
                      />
                    </g>
                  )}
                  {this.state.showTitleMessage && (
                    <UnitText
                      text="Changing Field-Of-View..."
                      x={width / 2}
                      y={80}
                      fontSize={width / 20}
                      style={{
                        fill: 'aqua',
                        width: '100%',
                        fontFamily: 'BrandonGrotesque-Black',
                      }}
                    />
                  )}
                  {activeInstrumentID && previousInstrumentID && isGridActive && (
                    <FadeSVG isHidden={!isTransitioningTelescope}>
                      <FieldOfView
                        activeInstrumentID={activeInstrumentID}
                        previousInstrumentID={previousInstrumentID}
                        telescopeId={this.state.telescopeId}
                        tickSpacing={tickSpacing}
                        canvasWidth={width}
                        currentZoomIn={this.currentZoomInTransition}
                        currentZoomOut={this.currentZoomOutTransition}
                        stroke={this.state.transitionStrokeColor}
                        showArrows={this.state.showArrows}
                      />
                    </FadeSVG>
                  )}
                  {isGridActive && (
                    <Fragment>
                      <TelescopeFrame
                        isGridVisible={isTransitioningTelescope}
                        isScaleVisible={
                          !isTransitioningTelescope &&
                          !this.state.showTitleMessage
                        }
                        horizontalResolution={horizontalResolution}
                        verticalResolution={verticalResolution}
                        increment={increment}
                        length={width}
                      />

                      <FadeSVG
                        isHidden={
                          isTransitioningTelescope ||
                          this.state.showTitleMessage
                        }
                      >
                        <Scale
                          dimension={width}
                          scale={
                            tickSpacing *
                            activeInstrument.directionMarkerLengthArcMinutes
                          }
                          scaleText={
                            activeInstrument.directionMarkerLengthArcMinutes
                          }
                          style={{ stroke: 'aqua' }}
                        />

                        <UnitText
                          text="arcminutes"
                          x={midPoint}
                          y={40}
                          style={{ letterSpacing: arcMinuteLabelLetterSpacing }}
                        />

                        <UnitText
                          text="arcminutes"
                          x={-midPoint}
                          y={width - 40}
                          style={{
                            letterSpacing: arcMinuteLabelLetterSpacing,
                            transform: 'rotate(-90)',
                          }}
                        />
                      </FadeSVG>
                    </Fragment>
                  )}
                </FadeSVG>
                {isModalActive && (
                  <Modal
                    show
                    size="lg"
                    dialogClassName="telescope-modal"
                    centered
                    onHide={this.onHideModal}
                  >
                    <h3
                      style={{
                        color: 'white',
                        marginTop: '-60px',
                        marginBottom: '15px',
                      }}
                    >
                      {missionTitle || 'No mission available'}
                    </h3>
                    <Telescope {...this.props} disableFullscreen />
                  </Modal>
                )}

                {transitionScale && (
                  <HowBig
                    dimension={width}
                    referenceObjectScale={missionMetaData.referenceObjectScale}
                    domain={missionMetaData.domain}
                    targetObjectScale={missionMetaData.targetObjectScale}
                    targetObjectURL={missionMetaData.targetObjectURL}
                    targetObjectName={missionMetaData.targetObjectName}
                    onComplete={this.handleCompleteHowBigAnimation}
                  />
                )}
              </svg>

              <style jsx>
                {`
                  .portal {
                    width: 100%;
                    overflow: hidden;
                    background: none;
                    position: relative;
                  }

                  .portal:before {
                    content: '';
                    padding-top: 100%;
                    float: left;
                  }

                  .portal :global(.telescope-svg) {
                    z-index: 5;
                  }

                  .portal :global(.telescope-float-menu) {
                    position: absolute;
                    top: 40px;
                    left: 26px;
                    z-index: 3000;
                  }

                  :global(.telescope-modal) {
                    max-width: 90vh;
                    width: 100%;
                  }

                  svg {
                    position: absolute;
                    left: 0;
                    right: 0;
                    bottom: 0;
                    top: 0;
                    width: 100%;
                    height: 100%;
                  }
                `}
              </style>
            </div>
          </div>
        )}
      </Measure>
    );
  }
}

const mapStateToProps = ({ starshareCamera }) => ({
  previousInstrumentId: starshareCamera.previousInstrumentId,
});

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      setPreviousInstrument,
    },
    dispatch
  );

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Telescope);
