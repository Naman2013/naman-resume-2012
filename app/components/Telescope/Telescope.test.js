import React from 'react';
import { shallow, mount } from 'enzyme';
import Telescope from './Telescope';
import telescopeConfig from './telescopeConfig';
import FAUX_MISSIONS from 'content/fauxMissions';

describe('Telescope interface', () => {
  const activeTelescope = telescopeConfig.CANARY_ONE_HALF_METER;
  const previousTelescope = telescopeConfig.CANARY_TWO_WIDE_FIELD;

  const shallowTelescope = shallow(<Telescope
    activeInstrumentID={activeTelescope.instrumentID}
    previousInstrumentID={previousTelescope.instrumentID}
    increment={5}
  />);

  const mountedTelescope = mount(<Telescope
    activeInstrumentID={activeTelescope.instrumentID}
    previousInstrumentID={previousTelescope.instrumentID}
    increment={5}
  />);

  describe('initialization', () => {
    it('should render correctly', () => {
      expect(shallowTelescope).toMatchSnapshot();
    });

    describe('`props`', () => {
      it('initializes with a `missionTargetID` of 0', () => {
        const initializingMissionMeta = { missionTargetID: 0 };
        expect(mountedTelescope.props().missionMetaData).toEqual(initializingMissionMeta);
      });
    });

    describe('`state`', () => {
      it('sets `awaitingMission` when `props.missionMetaData.missionTargetID` is `0`', () => {
        expect(mountedTelescope.state().awaitingMission).toEqual(true);
      });

      it('initializes `transitionScale` to `false`', () => {
        expect(mountedTelescope.state().transitionScale).toEqual(false);
      });
    });
  });

  describe('when updating with new mission information', () => {
    const { nonMission, scaleUp, scaleDown } = FAUX_MISSIONS;

    beforeEach(() => {
      mountedTelescope.setProps({ missionMetaData: nonMission });
    });

    describe('when the incoming ID is `0`', () => {
      beforeEach(() => {
        mountedTelescope.setProps({ missionMetaData: nonMission });
      });

      it('sets `awaitingMission` to `true`', () => {
        expect(mountedTelescope.state().awaitingMission).toEqual(true);
      });

      it('sets `transitionScale` to `false`', () => {
        expect(mountedTelescope.state().transitionScale).toEqual(false);
      });
    });

    describe('when the incoming ID is non-`0`', () => {
      beforeEach(() => {
        mountedTelescope.setProps({ missionMetaData: scaleUp });
      });

      it('sets `awaitingMission` to `false` when a real mission arrives', () => {
        expect(mountedTelescope.state().awaitingMission).toEqual(false);
      });

      describe('when a new `missionTargetID` arrives', () => {
        it('sets `transitionScale` to `true` when first non-`0` appears', () => {
          expect(mountedTelescope.state().transitionScale).toEqual(true);
        });

        it('sets `transitionScale` to `true` when new `missionTargetID` arrives', () => {
          mountedTelescope.setProps({ missionMetaData: scaleDown });
          expect(mountedTelescope.state().transitionScale).toEqual(true);
        });

        it('`transitionScale` remains `false` when seeing the same mission appear twice', () => {
          mountedTelescope.setProps({ missionMetaData: scaleDown });
          mountedTelescope.instance().handleCompleteHowBigAnimation();
          mountedTelescope.setProps({ missionMetaData: scaleDown });
          expect(mountedTelescope.state().transitionScale).toEqual(false);
        });
      });
    });

    describe('when `HowBig` animation is completed', () => {
      beforeEach(() => {
        mountedTelescope.instance().handleCompleteHowBigAnimation();
      });

      it('sets `transitionScale` to `false`', () => {
        expect(mountedTelescope.state().transitionScale).toEqual(false);
      });
    });
  });
});
