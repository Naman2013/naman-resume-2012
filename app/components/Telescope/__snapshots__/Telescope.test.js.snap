// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Telescope interface initialization should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Telescope
    activeInstrumentID="76de934d-7909-11e6-a635-0eb2b1774883"
    increment={5}
    missionMetaData={
      Object {
        "missionTargetID": 0,
      }
    }
    previousInstrumentID="8032dfaf-7909-11e6-a635-0eb2b1774883"
    render={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "bounds": true,
      "children": [Function],
      "onResize": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "bounds": true,
        "children": [Function],
        "onResize": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
